libfoundation: Add MCSpan class for wrapping pointer + length pairs

peter-b opened the pull request:
There are many places in the engine where (pointer, length) pairs are
handled separately.  Some bugs have occurred where they get separated
-- for example, when the pointer gets erroneously passed to a function
that expects to receive a null-terminated array rather than an array
with an extrinsic length.

The `MCSpan` class is inspired by the `span` class recommended by the
C++ Core Guidelines [1] and the `gsl::span` class included in
Microsoft's Guideline Support Library [2].  It is intended to be a
zero-cost abstraction that combines pointer + length into a single
class, and defines safe, checked access and subscripting operations.

It's not quite right yet, because to be fully zero-cost it requires
all of its methods to be declared `constexpr` and not all of the
compilers we use support that yet.

[1] https://github.com/isocpp/CppCoreGuidelines
[2] https://github.com/Microsoft/GSL

This PR demonstrates how easy it is to convert from (pointer, length) pairs to `MCSpan` using the `MCU_strtol()` function in libfoundation.

peter-b committed:
libfoundation: Add MCSpan class for wrapping pointer+length pairs.

There are many places in the engine where (pointer, length) pairs are
handled separately.  Some bugs have occurred where they get separated
-- for example, when the pointer gets erroneously passed to a function
that expects to receive a null-terminated array rather than an array
with an extrinsic length.

The `MCSpan` class is inspired by the `span` class recommended by the
C++ Core Guidelines [1] and the `gsl::span` class included in
Microsoft's Guideline Support Library [2].  It is intended to be a
zero-cost abstraction that combines pointer + length into a single
class, and defines safe, checked access and subscripting operations.

It's not quite right yet, because to be fully zero-cost it requires
all of its methods to be declared `constexpr` and not all of the
compilers we use support that yet.

[1] https://github.com/isocpp/CppCoreGuidelines
[2] https://github.com/Microsoft/GSL

peter-b committed:
libfoundation: Make MCU_strtol() operate on a span.

Demonstrates how straightforward it can be to convert code from
pointer + length representation to `MCSpan` representation.

livecodefraser commented:
@livecode-vulcan review okay 046cb15

livecode-vulcan commented:
💙 review by @livecodefraser ok 046cb15293475394562e03a5ca954b05d4ade762

livecode-vulcan commented:
😞 test failure 046cb15293475394562e03a5ca954b05d4ade762
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/459)
- try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/464)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/459)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/459)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/465)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/459)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/459)

peter-b committed:
libfoundation: Use decltype(nullptr), not std::nullptr_t

`std::nullptr_t` isn't defined in the `<cstddef>` header of every libc++ that we currently use.

Using `decltype(nullptr)` requires no library support because `nullptr` is a keyword.

livecodefraser commented:
@livecode-vulcan review okay c5a3409

livecode-vulcan commented:
💙 review by @livecodefraser ok c5a3409233dcbec61638d0fabde55b10a4cc0484

livecode-vulcan commented:
😞 test failure c5a3409233dcbec61638d0fabde55b10a4cc0484
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/460)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/465)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/460)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/460)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/466)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/460)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/460)

peter-b committed:
libfoundation: Don't use "using" to define type aliases.

MSVC 2010 can't cope with C++11-stype type definitions with the
`using` keyword.  Use `typedef` instead.

montegoulding commented:
@livecode-vulcan review ok 1858521

livecode-vulcan commented:
💙 review by @montegoulding ok 185852158eeb109f869e0e57d278ac5098038ae0

livecode-vulcan commented:
😎 test success 185852158eeb109f869e0e57d278ac5098038ae0
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/463)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/468)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/463)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/463)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/469)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/463)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/463)

peter-b committed:
libfoundation: MCU_strtol() should take r_remainder as reference.

Fixes a bug introduced in commit 046cb15.

livecodeali commented:
@livecode-vulcan review ok b87e29b

livecode-vulcan commented:
💙 review by @livecodeali ok b87e29b27482b78222b655951353be131d911b67

livecode-vulcan commented:
😎 test success b87e29b27482b78222b655951353be131d911b67
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/484)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/489)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/484)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/484)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/490)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/484)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/484)
